### ISSUE TITLE
_build_account_dic returns empty dictionary

### ISSUE THREAD

**Issue by githubuser3000** [grindsa/dkb-robo#47]
Hi, 

very recently the robo stopped working. Login works, but account_dic is empty. I stepped into the problem and found out, that productGroups are empty in the data, that is sent by DKB:

`{'product_display': {'data': [{'type': 'productDisplaySettings', 'id': '####################', 'attributes': {'portfolio': 'dkb', 'productSettings': {}, 'productGroups': {}, 'clientTypeSpecificSettings': {}}, 'links': {'self': 'https://api.developer.dkb.de/config/users/me/product-display-settings/####################'}}], 'included': []}`

But _build_account_dic needs this data to build the account dic.
The rest of the portfolio_dic looks totally fine.
Without the productGroups, no account dic, no transaction_urls and no chance to get_transactions. 

Tested versions 0.23.2 and 0.23.1, legacy_login=False

Anyone else has this issue? Is it really necessary to go via productGroups? It looks, like the account_dic could also be build without it.

Best regards

**Comment by grindsa** [comment]
Thank you for your explanation. Although I am not able to replicate your issue, I understand what's going wrong. I have an idea how to work around a scenario in which the productGroups attribute is empty. Give me a bid of time to work on a fix...

**Comment by grindsa** [comment]
hi again, a potential fix got slotted into the [devel branch](https://github.com/grindsa/dkb-robo/tree/devel) can you please check if this solves your issue?

**Comment by githubuser3000** [comment]
Hey, thanks for the update.
I tested it and account dictionary is now filled correctly. 

However, there are many things, that has been changed by dkb recently and has to be adapted in source code now.
Some of them are e.g.:
in account dic:
- IBANs are now represented without spaces
- date format has changed from %d.%m.%Y to %Y-%m-%d

in transactions:
- date format has also been changed
- "mandatreferenz" attribute not available anymore
- "customerreferenz" attribute not always available in each transaction

This only as hint for everyone observing the same changes, since it can't really be changed in the robo.

However, there is one issue remaining:
It seems, like dkb has changed the logic for account balance. The balance, that online banking shows and that is also returned by account dic now sometimes also includes transactions from the future (happens usually at weekends for transactions on next Monday).
get_transactions does not allow date_to in the future (due to validate_dates). This causes the situation, that account balance in account dic shows a value, that can't be reproduced by all transactions returned by get_transactions. This always causes a validation error in my code in this case.
Solution:
Simply allow date_to in the future, if this doesn't cause other issues.

**Comment by grindsa** [comment]
Hi,

Sorry for not responding earlier.  When building the fix for the issue you reported I did encounter further issues during regression, and it took a bid of time to fix all of them.

You are absolutely right it makes sense to create a list of "breaking changes when switching to the new API".  I am thinking to update #42 along with migrating the different methods to the new API; if you have a better idea let me know.

I will implement the "if date_to_uts > now_uts:" change you mentioned next and package a new release tomorrow or the day after.

Sorry again for keeping you waiting and thanks a lot for providing feedback.

**Comment by githubuser3000** [comment]
No problem, there is no hurry.

The unpleasant observation is, that the breaking changes didn't came with the new API right away. I used the the new one already a while without the problems and then suddenly it changed. And even worse, I found out, that other accounts do not yet have rolled out the breaking changes in the new API. It might happen suddenly in the background for each account individually.

I try to help as best as I can. And thanks a lot for all your work on this library. Its such a great tool!
